c7n/resources/iam.py

```diff
@@ -37,10 +37,243 @@ class Role(QueryResourceManager):
+                results.append(result)
+        for result in self.scan_ec2_roles(False):
+            if result not in results:
+                results.append(result)
+        return results
+
+    def get_profile_role_arn(self, resource):
+        results = []
+        client = local_session(self.manager.session_factory).client('iam')
+        for iprofile in client.list_instance_profiles()['InstanceProfiles']:
+            if iprofile['Arn'] == resource:
+                profile = client.get_instance_profile(
+                    InstanceProfileName=iprofile['InstanceProfileName'])
+                if 'Roles' not in profile:
+                    return []
+                for role in profile['Roles']:
+                    results.append(role['Arn'])
+        return results
+
+    def scan_lambda_roles(self):
+        results = []
+        client = local_session(self.manager.session_factory).client('lambda')
+        fxs = client.list_functions()['Functions']
+        for fx in fxs:
+            if 'Role' not in fx:
+                continue
+            results.append(fx['Role'])
+        return results
+
+    def scan_ecs_roles(self):
+        results = []
+        client = local_session(self.manager.session_factory).client('ecs')
+        for cluster in client.describe_clusters()['clusters']:
+            svcs = client.list_services(cluster=cluster)['serviceArns']
+            for svc in client.describe_services(cluster=cluster,services=svcs)['services']:
+                if 'roleArn' not in svc:
+                    continue
+                results.append(svc['roleArn'])
+        return results
+
+    def scan_asg_roles(self, rolearn):
+        results = []
+        client = local_session(self.manager.session_factory).client('autoscaling')
+        asgs = client.describe_launch_configurations()['LaunchConfigurations']
+        for asg in asgs:
+            if 'IamInstanceProfile' not in asg:
+                continue
+            if rolearn:
+                for arn in self.get_profile_role_arn(asg['IamInstanceProfile']):
+                    results.append(arn)
+            else:
+                results.append(asg['IamInstanceProfile'])
+        return results
+
+    def scan_ec2_roles(self, rolearn):
+        results = []
+        client = local_session(self.manager.session_factory).client('ec2')
+        instances = client.describe_instances()['Reservations']
+        for instance in instances:
+            for ec2 in instance['Instances']:
+                if 'IamInstanceProfile' not in ec2:
+                    continue
+                if rolearn:
+                    for arn in self.get_profile_role_arn(ec2['IamInstanceProfile']['Arn']):
+                        results.append(arn)
+                else:
+                    results.append(ec2['IamInstanceProfile']['Arn'])
+        return results
+
+
+###################
+#    IAM Roles    #
+###################
+
+
+@Role.filter_registry.register('inuse')
```

Comment:
rename for consistency, 'used' and 'unused' are the vocabulary patterns we've established for this.